src/testing/slim_check/testable.lean

```diff
@@ -0,0 +1,490 @@
+def give_up (x : ℕ) : test_result p → test_result p
+ | (success (psum.inl ())) := gave_up x
+ | (success (psum.inr p))  := success (psum.inr p)
+ | (gave_up n) := gave_up (n+x)
+ | (failure Hce xs) := failure Hce xs
+
+variable (p)
+
+variable [testable p]
+
+/-- configuration for testing a property -/
+@[derive [has_reflect, inhabited]]
+structure slim_check_cfg :=
+(num_inst : ℕ := 100) -- number of examples
+(max_size : ℕ := 100) -- final size argument
+
+/-- Try `n` times to find a counter-example for `p` -/
+def testable.run_suite_aux (cfg : slim_check_cfg) : test_result p → ℕ → rand (test_result p)
+ | r 0 := return r
+ | r (succ n) :=
+do let size := (cfg.num_inst - n - 1) * cfg.max_size / cfg.num_inst,
+   x ← retry ( (testable.run p tt).run ⟨ size ⟩) 10,
+   match x with
+    | (success (psum.inl ())) := testable.run_suite_aux r n
+    | (success (psum.inr Hp)) := return $ success (psum.inr Hp)
+    | (failure Hce xs) := return (failure Hce xs)
+    | (gave_up g) := testable.run_suite_aux (give_up g r) n
+   end
+
+/-- Try to find a counter-example of `p` -/
+def testable.run_suite (cfg : slim_check_cfg := {}) : rand (test_result p) :=
+testable.run_suite_aux p cfg (success $ psum.inl ()) cfg.num_inst
+
+/-- Run a test suite for `p` in `io` -/
+def testable.check' (cfg : slim_check_cfg := {}) : io (test_result p) :=
+io.run_rand (testable.run_suite p cfg)
+
+namespace tactic
+
+open tactic expr
+
+/-!
+## Decorations
+
+Instances of `testable` use `named_binder` as a decoration on
+propositions in order to access the name of bound variables, as in
+`named_binder (some "x") (forall x, x < y)`.  This helps the
+`testable` instances create useful error messages where variables
+are matched with values that falsify a given proposition.
+
+The following functions help support the gadget so that the user does
+not have to put them in themselves.
+-/
+
+/-- `add_existential_decorations p` adds `a `named_binder` annotation at the
+root of `p` if `p` is an existential quantification -/
+meta def add_existential_decorations : expr → expr
+| e@`(@Exists %%α %%(lam n bi d b)) :=
+  let n := to_string n in
+  const ``named_binder [] (const ``option.some [level.zero] (`(string) : expr) (`(n) : expr)) e
+| e := e
+
+/-- Traverse the syntax of a proposition to find universal quantifiers
+and existential quantifiers and add `named_binder` annotations next to
+them. -/
+meta def add_decorations : expr → expr | e :=
+e.replace $ λ e _,
+  match e with
+  | (pi n bi d b) :=
+    let n := to_string n in
+    some $ const ``named_binder [] (const ``option.some [level.zero] (`(string) : expr) (reflect n : expr)) (pi n bi (add_existential_decorations d) (add_decorations b))
+  | e := none
+  end
+
+/-- `decorations_of p` is used as a hint to `mk_decorations` to specify
+that the goal should be satisfied with a proposition equivalent to `p`
+with added annotations. -/
+@[reducible, nolint unused_arguments]
+def decorations_of (p : Prop) := Prop
```

Comment:
I'm still trying to sort out exactly what's going on with these decorations, but this seems odd to have in the `tactic` namespace. (Note that you're still in `slim_check` too. If you wanted `mk_decorations` in `_root_.tactic`, it isn't there right now.)

Reply:
Basically, the instance `testable (named_binder (some "x") (∀ x, p x))` generates samples for `x` and, when the test fails, adds `x := 37` to the error message that will be printed to the user. `testable` is pure code and does not have access to the syntax tree (esp. the name of bound variables) of `∀ x, p x` so we use a gadget to make the name "x" available.